### PR TITLE
Adjust builder to validate that JSON in Dockerfiles are arrays of strings and nothing else to match how we describe them to people (and what all our existing tests already assumed)

### PR DESCRIPTION
**Pull request by tianon**
This also adds more tests to help verify this, including unicode and nonprintable characters (hence the commit switching to `strconv.Quote`).

As a bonus, this fixes a subtle bug where `[]` was turned into `[""]` and then turned back into `[]` (and thus `[""]` was impossible to actually round-trip correctly in a Dockerfile).

In the process, I replaced the custom `QuoteString` function with the equivalent (but with more special cased characters) `strconv.Quote` in `Node.Dump()`, and simplified the builder `TestTestData` a little by using `ioutil.ReadFile` directly instead of `os.Open` followed immediately by `ioutil.ReadAll`.

### CODE REVIEW

File: builder/parser/line_parsers.go [hunk 1]
```diff
@@ -209,34 +208,28 @@ func parseString(rest string) (*Node, map[string]bool, error) {
...
 	if err := json.Unmarshal([]byte(rest), &myJson); err != nil {
 		return nil, nil, err
 	}
 
+	var top, prev *Node
 	for _, str := range myJson {
 		switch str.(type) {
 		case string:
-		case float64:
-			str = strconv.FormatFloat(str.(float64), 'G', -1, 64)
```
Comment (LK4D4) [inline builder/parser/line_parsers.go]: THIS IS BREAKING CHANGE!!!111!!

Comment (tianon) [inline builder/parser/line_parsers.go]: No tests for it, no guarantee of non breakage.

Comment (tianon) [inline builder/parser/line_parsers.go]: Also, we never describe it this way.  Can you find me `Dockerfile`s in the wild that use it legitimately?  Is there actually a case where a `float64` is _preferred_ to a string?  For reals?

File: builder/parser/parser_test.go [hunk 2]
```diff
@@ -54,18 +54,14 @@ func TestTestData(t *testing.T) {
 		if err != nil {
 			t.Fatalf("Dockerfile missing for %s: %s", dir.Name(), err.Error())
 		}
-
-		rf, err := os.Open(resultfile)
-		if err != nil {
-			t.Fatalf("Result file missing for %s: %s", dir.Name(), err.Error())
-		}
+		defer df.Close()
```
Comment (jfrazelle) [inline builder/parser/parser_test.go]: lol what? ^

File: builder/parser/line_parsers.go (hunk 1 above)
Comment (jfrazelle) [inline builder/parser/line_parsers.go]: it really makes no sense for a cmd or entrypoint

Comment (tianon) [inline builder/parser/line_parsers.go]: or a `RUN` line either :wink:

File: builder/parser/line_parsers.go [hunk 3]
```diff
@@ -209,34 +208,28 @@ func parseString(rest string) (*Node, map[string]bool, error) {
...
 // parseJSON converts JSON arrays to an AST.
 func parseJSON(rest string) (*Node, map[string]bool, error) {
-	var (
-		myJson   []interface{}
-		next     = &Node{}
-		orignext = next
-		prevnode = next
-	)
-
+	var myJson []interface{}
```
Comment (LK4D4) [inline builder/parser/line_parsers.go]: Why not just `[]string`? I tested, Unmarshal returns nice error if there is something apart from strings in array.

Comment (tianon) [inline builder/parser/line_parsers.go]: I actually started with that:
```irc
16:59:08 <tianon> heh, we actually get a failing test because the nested arrays are no longer an error, but actually get passed directly to the shell as-is
17:07:08 <tianon> so we _have_ established that if something parses as valid JSON array syntax, it must be valid Dockerfile syntax (and is not passed to the shell)
```

Comment (tianon) [inline builder/parser/line_parsers.go]: (because if something fails to parse here, it actually just gets passed to the shell instead, unless it's this one specific special error)

File: builder/parser/line_parsers.go (hunk 1 above)
Comment (LK4D4) [inline builder/parser/line_parsers.go]: Everywhere in docs this is just "JSON Array". Also I am +100 on this breaking change. As I was +100 on breaking change for USER+COPY, so don't take my words too serious.

File: builder/parser/parser_test.go (hunk 2 above)
Comment (LK4D4) [inline builder/parser/parser_test.go]: Moved from end of test to defer I think

File: builder/parser/line_parsers.go (hunk 1 above)
Comment (tianon) [inline builder/parser/line_parsers.go]: If the docs maintainers want me to, I can go on a docs dive replacing `JSON array` with `JSON array of strings`... :+1: :wink:

File: builder/parser/line_parsers.go (hunk 3 above)
Comment (LK4D4) [inline builder/parser/line_parsers.go]: I like `json.UnmarshalTypeError` more than `errDockerfileNotStringArray`.

Comment (tianon) [inline builder/parser/line_parsers.go]: I agree, but that won't work.  The fact that it doesn't parse as valid JSON is what tells us it's supposed to be a shell string. :wink:

Comment (tianon) [inline builder/parser/line_parsers.go]: ie, `echo hi > /some/file` isn't valid JSON :smile:

File: builder/parser/parser_test.go (hunk 2 above)
Comment (jfrazelle) [inline builder/parser/parser_test.go]: I was talking about the `os.Open()` to `ioutil.ReadAll()` lol ;)

File: builder/parser/line_parsers.go (hunk 3 above)
Comment (LK4D4) [inline builder/parser/line_parsers.go]: @tianon This is `json.SyntaxError` not `json.UnmarshalTypeError`

Comment (tianon) [inline builder/parser/line_parsers.go]: Ahh, genius.  Will refactor.

Comment (tianon) [inline builder/parser/line_parsers.go]: Except, we still have a problem.  `RUN true` is valid JSON, but needs to be passed to the shell.

Comment (tianon) [inline builder/parser/line_parsers.go]: Sorry @LK4D4, I don't think we can win here, unless you have something up your sleeve to tell `json` to only parse arrays...  :wink:

Comment (LK4D4) [inline builder/parser/line_parsers.go]: Json is sucks. As I said before :D

Comment (LK4D4) [inline builder/parser/line_parsers.go]: Also, can't be some conditions in bash vailid json arrays? I remember that there is some `[]` syntax.

Comment (tianon) [inline builder/parser/line_parsers.go]: Yes, but we've already broken any chance at those, and we have tests to validate that. :cry:

Comment (tianon) [inline builder/parser/line_parsers.go]: Generally though, `RUN [ some-test-here ]` _will_ work, as long as it's not valid JSON syntax.

Comment (tianon) [inline builder/parser/line_parsers.go]: (which is still super useful)

File: builder/parser/line_parsers.go [hunk 4]
```diff
@@ -10,13 +10,12 @@ import (
...
 	"errors"
 	"fmt"
-	"strconv"
 	"strings"
 	"unicode"
 )
 
 var (
-	errDockerfileJSONNesting = errors.New("You may not nest arrays in Dockerfile statements.")
+	errDockerfileNotStringArray = errors.New("When using JSON array syntax, arrays must be comprised of strings only.")
```
Comment (thaJeztah) [inline builder/parser/line_parsers.go]: "must be comprised of"; Perhaps something simpler, like "arrays can only contain strings / string-values". Trying to not use "comprised" here, because that may be difficult to understand for non-English users (not sure)

**Comment by erikh** [comment]
You should probably make Node values take string only now instead of interface. Should speed the builder up.

File: builder/parser/line_parsers.go (hunk 4 above)
Comment (tianon) [inline builder/parser/line_parsers.go]: This phrase is only a 6.8 on the Flesch-Kincaid Grade Level for
Readability. ;)

https://github.com/docker/docker/blob/master/docs/sources/contributing/docs_style-guide.md#prose-style

**Comment by tianon** [comment]
I'm not sure what you mean?
https://github.com/docker/docker/blob/b2ab733c993ee6124c164defceac6996206d69fa/builder/parser/parser.go#L25

Aren't they already?

File: builder/parser/line_parsers.go (hunk 4 above)
Comment (thaJeztah) [inline builder/parser/line_parsers.go]: :o) Hm, okay.. might be just me then? LOL according to that "bye" is 330% more readable than "hello" :smile:

**Comment by erikh** [comment]
sorry I misread some of the above code, it's been a while since I was in here.

File: builder/parser/line_parsers.go [hunk 5]
```diff
@@ -209,34 +208,28 @@ func parseString(rest string) (*Node, map[string]bool, error) {
...
-	)
-
+	var myJson []interface{}
 	if err := json.Unmarshal([]byte(rest), &myJson); err != nil {
 		return nil, nil, err
 	}
 
+	var top, prev *Node
 	for _, str := range myJson {
 		switch str.(type) {
```
Comment (icecrime) [inline builder/parser/line_parsers.go]: Do we need a type switch rather than a single type assertion? I don't think it changes anything besides readability, but I'd find `if s, ok := str.(string); !ok {` easier (and avoids casting twice).

File: builder/parser/line_parsers.go (hunk 1 above)
Comment (SvenDowideit) [inline builder/parser/line_parsers.go]: so you mean you've broken `RUN ["echo", 3.14]` ? perhaps mention it in the release notes, someone will trip on it, but I too would expect it to be very few.

**Comment by jfrazelle** [comment]
LGTM but maybe just a single type assertion like @icecrime suggested :)

File: builder/parser/line_parsers.go (hunk 1 above)
Comment (tianon) [inline builder/parser/line_parsers.go]: We don't have a place where we currently aggregate "next version" release notes, right?

**Comment by tianon** [comment]
Updated type switch to be just a type assertion instead. :+1:

**Comment by erikh** [comment]
lgtm

**Comment by jfrazelle** [comment]
I think we are just waiting on @tiborvass

**Comment by jfrazelle** [comment]
@duglin feel free to review and merge at your convenience :)

**Comment by duglin** [comment]
What should:  RUN [ "-e foo" ]   
result in?  I get different results from the shell vs docker(this pr).
Should this PR have fixed this ambiguity?

File: builder/parser/parser.go [hunk 6]
```diff
@@ -85,10 +85,7 @@ func parseLine(line string) (string, *Node, error) {
 		return "", nil, err
 	}
 
-	if sexp.Value != "" || sexp.Next != nil || sexp.Children != nil {
-		node.Next = sexp
-	}
-
+	node.Next = sexp
```
Comment (tiborvass) [inline builder/parser/parser.go]: @tianon why this change?

Comment (tianon) [inline builder/parser/parser.go]: This `if` was here to check for the case of `[]` turning into `[""]` and turning it back into `[]`.  By fixing `parseJSON` to return `nil` (the linked-list representation of an empty list), this `if` is no longer required, so now `[]` and `[""]` can both round-trip successfully.

**Comment by tianon** [comment]
@duglin that's a valid JSON array of strings, and there's no way for us to detect that it should be parsed by the shell as-is instead, so that becomes the literal command `-e foo` (which very likely won't exist in `PATH`)

**Comment by tianon** [comment]
Which is the current behavior without this PR, too:
```console
$ docker build -
FROM busybox
RUN [ "-e foo" ]
Sending build context to Docker daemon 2.048 kB
Sending build context to Docker daemon 
Step 0 : FROM busybox
 ---> 4986bf8c1536
Step 1 : RUN -e foo
 ---> Running in 439e68f47660
/bin/sh: illegal option - 
INFO[0006] The command [/bin/sh -c -e foo] returned a non-zero code: 2 
```

**Comment by tianon** [comment]
Although it's strange that we're getting `/bin/sh` there when we shouldn't be...

**Comment by jfrazelle** [comment]
what was the conclusion here?

**Comment by LK4D4** [comment]
@jfrazelle that json sucks, but this PR has 2 l.g.t.m.s already :)

**Comment by thaJeztah** [comment]
Nice JSON related issue: https://github.com/docker/docker/issues/10097

**Comment by jfrazelle** [comment]
i dont know eriks was not all caps

**Comment by erikh** [comment]
IHERDAT
